### ISSUE TITLE
Typing a non-arriw function leads to some partially working result

### ISSUE THREAD

**Issue by ericmorand** [microsoft/TypeScript#56636]
### 🔎 Search Terms

type function
type non arrow function

### 🕗 Version & Regression Information

I didn't test outside of TS 5.2.2 but I assume it was not working before either.

### 💻 Code

```ts
type MyFunction = (foo: number) => string;

<MyFunction>function doSomething(foo) => {
    return `${foo}`;
}

doSomething(5);
```

### 🙁 Actual behavior

Here, both `foo` and the return type of `doSomething` are perfectlt inferred - returning a `string` would fail the compilation, just like using `foo` as anything else than a number.

But...the next instruction `doSomething(5);` throws a "TS2552: Cannot find name `doSomething`." which obviously makes the code useless.

### 🙂 Expected behavior

I'm not sure. Ideally I would expect that it works - i.e. that using `doSomething` remains possible - but I'm not sure it was ever intended to type non-arrow function at all.

What is strange though is that it seems to work just fine, until we try to actually use the function.

### Additional information about the issue

This is related to Rollup plugin development where plugin hooks are non-arrow function. It would be amazing if we could type them individually to organize the code better instead of having to put everything under an object definition typed as a `Plugin`, like this:

```ts
const plugin: Plugin = {
    transform(this, code, id) { // all parameters types are perfectly inferred
    }
};
```

With this syntax, obviously, the `transform` method type is correctly inferred but it would be cool if we were allowed to write it this way for better reusability:

```ts
<Plugin["transform"]>function transform(this, code, id) { // all parameters types are perfectly inferred
}

const plugin: Plugin = {
    transform
};
```

See https://github.com/rollup/rollup/issues/5277

**Comment by jcalz** [comment]
Duplicate of #14520. This is a missing feature, not a bug.

**Comment by MartinJohns** [comment]
Functions can be expressions as well, so TypeScript probably interprets it as an expression that is being type asserted, not a top-level function declaration.

```typescript
const test = <any>function foo() {}
```

**Comment by nmain** [comment]
> Functions can be expressions as well, so TypeScript probably interprets it as an expression that is being type asserted, not a top-level function declaration.

I believe that's what's happening, as you can see in the JS output it uses parens to maintain the function expression (and not function definition) syntax:

```javascript
(function doSomething(foo) { ... })
```

Given that JS output, the runtime JS behavior is understandable

**Comment by ericmorand** [comment]
So basically everything works as expected and it is perfectly possible to type a function - but it is considered as an expression that needs to be assigned to a variable to be referenced later. Did I understand things properly?

**Comment by nmain** [comment]
Yes, that's always been allowed (and will get a good contextual type and everything):

```typescript
var x: (value: string) => string = function bah(a) {
    return a + 4;
}
```

But it's a different thing than a function definition and behaves differently at runtime.

**Comment by ericmorand** [comment]
Many thanks. I was totally unaware of that.
